ocaml/typing/typecore.ml

```diff
@@ -5307,7 +5307,10 @@ and type_argument ?explanation ?recarg env (mode : expected_mode) sarg
         texp
       end else begin
       unify_exp env {texp with exp_type = ty_fun} ty_expected;
-      if args = [] then texp else
+      if args = [] then texp else begin
+      (* In this case, we're allocating a new closure, so [sarg] needs
+         to be valid at [mode_subcomponent mode], not just [mode] *)
+      submode ~loc:sarg.pexp_loc ~env mode.mode (mode_subcomponent mode);
```

Comment:
I think this is too strict. It is checking that the expected mode could be used as a subcomponent of the expected mode. Whereas we want to check that the actual mode of the argument could be used as a subcomponent of the expected mode. I think that you will need to create a mode variable to use as the expected mode of the argument, and then either submode it against either `mode` or `(mode_subcomponent mode)` based on whether we decide to do eta-expansion.

Something like this probably fails currently:
```ocaml
let foo p (f : ?b:bool -> unit -> int) (local_ g : unit -> int) =
  if p then (f : unit -> int) else g
```

I think that we also need a call to `register_allocation` so that we allocate this closure locally if possible.